### DISCUSSION TITLE
How to programmatically create a new user?

### DISCUSSION THREAD

**Discussion by horacioj** [oqtane/oqtane.framework#1620] [Q&A]
Hello,

This is the case:

I would like to auto-register users after validating credentials against another authentication provider, but this is irrelevant to the problem I'm facing.

After the user successfully authenticates in the external system, I would like to create (if it doesn't already exists) new a user account in Oqtane to make it work as any regular user account. How to take care of keeping things in sync and secure is irrelevant to to the problem I'm facing.

Everything goes well until I do:

```
user = await UserService.AddUserAsync(user);
```
This returns `null` because (a) nobody is already logged in (this module acts as an alternate to the standard login module), and (b) `AllowRegistration` is disabled in the site.

In `\oqtane.framework\Oqtane.Server\Controllers\UserController.cs` , inside ` private async Task<User> CreateUser(User user)`, the first problem I'm facing is that of course `if (User.IsInRole(RoleNames.Admin))` is false because there is nobody logged in.

In addition, 

```
verified = false;
allowregistration = _sites.GetSite(user.SiteId).AllowRegistration;
```
This means the user will not be created as verified and it will not be created at all because `AllowRegistration` is disabled.

So, the question is, how can I programmatically create a new user without having to be running the site already logged in as an Admin?

Is there a way to temporarily change the security context for this to work? (For example, Sitecore allows doing something like this. i.e. run a piece of code impersonating someone else).

Said all of this, which options do I currently have to programmatically create a new user in Oqtane avoiding the limitations and security checks set in  `private async Task<User> CreateUser(User user)`? 

Should I copy & paste the code and create my own version or is there a better way (i.e. using the Oqtane APIs as they are)?

Thanks!

** EDIT **
I understand Oqtane's  server API is not the same as the DNN API, but I guess there should be a way to do this kind of things 'server side' outside the context of who si running the code?

**Comment by sbwalker** [comment]
Oqtane uses a client/server architecture and it's REST API is designed to ensure only authorized clients have access. There is no option to bypass these security requirements, as that would expose a security vulnerability. If you want to create an external client application that adds users I would suggest you create a flow which first establishes a security context and then calls the API. You can use the standard HTTPClient or a library such as RestSharp to do this.

**Reply by horacioj** [reply]
Thank you, @sbwalker. Now I understand, but I was unable to do any successful request to Oqtane's server API. 

On [this other discussion](https://github.com/oqtane/oqtane.framework/discussions/1078),  @ADefWebserver was facing similar problems.

I tried  `POST https://localhost:44336/1/api/User/login?setCookie=true&isPersistent=false'` and also `POST https://localhost:44336/api/User/login?setCookie=true&isPersistent=false'`, but I think the first one is correct? Because it is including the alias ID? 

Anyway, something is still not working, testing at `https://localhost:44336/swagger/index.html` works fine, but not from my code.
I'm doing by code exactly the same I see on the Swagger test page. Same querystring, same body, same headers.  The difference is the URL. In Swagger is `https://localhost:44336/api/User/login?setCookie=true&isPersistent=false`. This URL in my code fails with error 403. Adding the /1, fails with error 400

What could be wrong?

A brief example code using Flurl, RestSharp or whatever could be very helpful :)

**Reply by sbwalker** [reply]
In Oqtane 2.1 the framework was modified so that the server API uses the same URL pattern as the client - it no longer relies on an AliasId segment to resolve the tenant. So if the alias is domain.com then the server URL is domain.com/api. If the alias is domain.com/folder then the server URL is domain.com/folder/api. The user login method is a POST so you need to send the parameters appropriately.

**Reply by horacioj** [reply]
The error 400 happens when calling the Oqtane API from a controller (server side, i.e. module's "Server" project, a controller class under the /Controllers folder). 

It works fine when calling the Oqtane API from a service (client side, i.e. module's "Client" project, a class under /Services) 

NOTE: I'm running Oqtane on server mode ("Runtime": "Server",).
Either way, making API calls with an embed superuser/password from the client project doesn't look right.  Or perhaps there are things about Oqtane and Blazor that I still don't understand and I see a big security issue where there isn't any?

This (security concerns) is why I was trying to make the API calls from the server side (within a controller). Which perhaps, doesn't make sense? But otherwise, how could I use the Oqtane API (being a REST API)?

Is it normal that I get error 400 trying to make Oqtane API call from a controller ("Server" project)?

Is calling `​/api​/User​/login` the right way to authenticate? Because I'm not getting any cookies. 
Is there a header I need to set to make the session/context valid for a given user/password when making the `​/api​/User​/login`  call? Or should I create the session differently?

Example (using [Flurl](https://flurl.dev/)) that works on the "Client" project but fails (error 400) on the "Server" project:

```
var url = "https://localhost:44336/api/User";

var loginUrl = url
            .AppendPathSegment("login")
            .SetQueryParam("setCookie", "true")   // NOTE: tried also "false" with same results (got no cookies)
            .SetQueryParam("isPersistent", "false");

using (var session = new CookieSession(url))   //NOTE: tried using loginUrl for the session with same results (no cookies)
{
    // NOTE: This call works from the client project but throws a 400 from the server project
    //       In addition, loginResult doesn't include any cookie
    var loginResult = await loginUrl
        .WithHeader("accept", "text/plain")
        .WithHeader("Content-Type", "application/json")
        .PostJsonAsync(user);  //NOTE: user has a siteid + a  host username & password

    //NOTE: I get here the expected result (the host user's User object)
    var returnedUser = await loginResult.GetJsonAsync<User>();

    //NOTE: this (trying to get a user's object) will fail with error 403. I.e. the 'session' isn't working.
    var someUser = await url.AppendPathSegment("name")
                .AppendPathSegment("TNEUSERNAME")
                .GetJsonAsync<User>();
    }
}
```

**Reply by horacioj** [reply]
I'm afraid there is no way to use the Oqtane API as it is currently is (using cookie authentication). Am I right?

**Reply by sbwalker** [reply]
@horacioj you are trying to use a client/server application as if it were a server-based application - which is not going to produce good results. Note that regardless of whether you are running Oqtane on Blazor Server or Blazor WebAssembly it still uses a client/server architecture. In a client/server architecture the client and server are independent from one another and they need to communicate via standard protocols - in this case it is HTTP REST with security context shared via a claims cookie. The server application should not call REST APIs over HTTP - only a client should call REST APIs. The title of this thread is how to programmatically create a new user... but it sounds like what you are actually trying to do is create an alternate approach for authentication ( ie. login against another system ). Am I correct? This is going to be challenging and will involve a lot more than just programmatically creating a user in Oqtane.

**Reply by horacioj** [reply]
> Am I correct?

Yep...  I cannot wait until authentication providers are implemented. I'm almost there, though. Hacky... and copy & pasting code instead of invoking it (which could have been better, but there is no way of doing it). I think later (once alternative authentication methods are implemented in Oqtane) I could be able to remove the copy/pasted code and do clean API calls.

Anyway, the problem of invoking the Oqtane API from an Oqtane module but under a different security context (not the user logged in) is still a valid problem, right?

For example, what if in a module, logged in as a simple "Registered User" (or as an "Anonymous" user), for some reason it is needed to get a list of all roles in the portal to do or show something to the user? 
It cannot be done, because the logged-in user doesn't have the rights. Only "Administrators" can query roles.

How can this be solved? I think it is a valid, and possible common use case (i.e. query the API under a different security context).
I could not find any way to do the Oqtane authentication in a new (programmatically built) security context, at least with the current Cookie authentication.

**Reply by sbwalker** [reply]
Authentication providers will allow a user to authenticate against authorities other than the default .NET Core Identity service. This will make authentication more flexible but it will not change the authorization requirements for the public API. The public API for core services needs to be secure and only provide access to trusted users. Allowing access to untrusted users would expose a security vulnerability to the framework. 

Oqtane has taken the least privilege approach to security ie. it has locked down the API to the most secure level by default and only adjusted the restriction in those cases where an actual practical need has been identified. This is the recommended approach for building any secure system. And it is possible that some APIs are overly restrictive at this point... however adjustments will not be made unless there is a clear case and it can be proven that it does not compromise the integrity of the system. 

As an example, allowing an anonymous user to retrieve a list of the roles defined for a site would be an information disclosure vulnerability - as this type of information is privileged and should only be available to trusted users.

**Reply by artmedia** [reply]
A practical approach to programmatically create a new user could be something like this:

1) insert a Key (string) in appsettings.json
2) insert the same Key in your client Module.
3) when you need to create the new user from your client Module create a custom API to which you pass: Key, Username (of the new user), Email (of the new user), Name (of the new user).
4) in your server side custom API compare Key from the client with Key from appsettings.json, if they match you can programmatically create new user (copy/past few lines from Oqtane framework in your code).

@sbwalker, is this approach secure enough?

**Reply by sbwalker** [reply]
@artmedia if you are asking if this approach is secure enough for the core framework, my answer would definitely be No. Any "secret" stored client-side is not protected. This is true in any SPA application regardless of whether it is based on JavaScript or Blazor.

**Reply by horacioj** [reply]
> The public API for core services needs to be secure and only provide access to trusted users. 
> Allowing access to untrusted users would expose a security vulnerability to the framework.

I totally agree, @sbwalker 
The problem is, there is no way to authenticate to the API programmatically, which make it useful only for the Oqtane core itself. It cannot be utilized in modules. It isn't acting as a public API because there is no way to authenticate for using it. This is a big problem, I think.

**Reply by sbwalker** [reply]
@horacioj I have to admit that I am really struggling to understand your questions and comments in this discussion thread. I have created many external modules for Oqtane and never run into any difficulties. Perhaps the issue you are encountering is specific to your use case. I don't think I fully understand what you are trying to accomplish... so that may be contributing to my confusion.

**Reply by horacioj** [reply]
Sorry, never mind. The answer to the original question is " you can't  programmatically create a new user" unless the user using your module is logged-in as an Administrator. The thread was no longer about what the title says... So, better leave it as it is.

**Reply by sbwalker** [reply]
If we focus on programmatically creating a new user from an Oqtane module there are a few possible options:

1. Call the Oqtane User API. In order to do this the logged in user must be a site administrator.
2. Enable "Allow User Registration" in Site Settings. This will allow anyone to create a user account using the Oqtane User API.
3. Expose your own API which contains a method that has more relaxed security constraints than the Oqtane API. Within this method you can implement whatever code you want. Please note that by taking this approach you may be exposing a security vulnerability for the site as you are effectively bypassing the security restrictions for the core framework - but that is ultimately your responsibility.

**Reply by horacioj** [reply]
Thank you, @sbwalker. The options are clear. 

Regarding the 3rd. option, (in my case, the module's API method taking care of creating a user is secured by a different external authentication), something like creating a user should be re-coded as there is no way to use the Oqtane logic (API) already created for this (`public async Task<User> Post([FromBody] User user)` /  `CreateUser(User user)` at `\Oqtane.Server\Controllers\UserController.cs`).

**Comment by W6HBR** [comment]
Hate to bring back an old topic, but I'm struggling with this similar problem.

I need non-admin users assigned to a custom role (e.g. "Profile Editors") to be able to create/edit users and their profile properties (about 40 profile settings fields), but not be a full site administrator.

My first thought was to use copies of the Users module Add/Edit/Roles screens in my own module, but then realized that the API security still wouldn't allow this to function.

I'm looking for a little more guidance than just #3 above from Shaun.  Seems that some granular assignment of rights via roles could benefit many orgs/users as having an all or none access via Administrator is just too restrictive.

As a work-around, how could I have a regular user call the built-in user add/edit screens via impersonation with an Admin user credential?

**Reply by sbwalker** [reply]
No worries about bringing up this topic again... clearly it is still a challenge. Oqtane is a modular application framework which means that flexibility is extremely important. It uses a client/server architecture which means that front-end is fully separated from the back-end. The back-end functionality is exposed via an API and currently that API uses some security conventions which are intended to safeguard the system from inappropriate use - however to accomplish this, the implementation lacks flexibility.  Specifically, many of the API methods include static role specifications for authorization (ie. a user must be in the Administrators role in order to perform an action). 

The optimal way to make the API more flexible would be to implement dynamic authorization policies. This would allow the Administrator of a site the ability to specify which roles or users are allowed to access a specific API. Most likely this would need to be split into Read/Write operations and in order to adhere to the "headless" approach, the permissions would need to be maintained independently from the current View/Edit permissions which are specific to the client UI. The separation would provide API management capabilities for other types of native client applications.

This is not a small amount of effort as it would affect the entire framework API layer. It would need to be implemented in a consistent manner across all APIs with a centralized UI for management. And it would need to support third party module APIs as well. Luckily, the existing permissions model in Oqtane was developed in flexible manner so it could support this use case.

**Reply by W6HBR** [reply]
As much as I hate making changes to the core framework, I solved this issue with a few small changes to UserController.cs, UserRoleController.cs and SettingController.cs.  @sbwalker - Your comment about the API methods including static role specifications gave me an idea to extend that logic with two additional static roles of "Profile Editors" and "Role Editors".

This now allows me to assign users to these roles for the purpose of maintaining site users without being a full site **Administrator**.

Example changes to UserRoleController.cs:
`[Authorize(Roles = RoleNames.Admin + ",Role Editors")]` 

Example changes to UserController.cs:
`[Authorize(Roles = RoleNames.Admin + ",Profile Editors")]` 
and
`if (!User.IsInRole("Profile Editors") && !User.IsInRole(RoleNames.Admin) && User.Identity.Name?.ToLower() != user.Username.ToLower())`

Example change to SettingController.cs:
`case EntityNames.User:`
`      filter = !User.IsInRole("Profile Editors") && !User.IsInRole(RoleNames.Admin) && _userPermissions.GetUser(User).UserId != entityId;`
